internal/nearcache/record_store.go

```diff
@@ -116,6 +163,79 @@ func (rs *RecordStore) Invalidate(key interface{}) {
 	rs.incrementInvalidationRequests()
 }
 
+func (rs *RecordStore) doEviction() bool {
+	// port of: com.hazelcast.internal.nearcache.impl.store.AbstractNearCacheRecordStore#doEviction
+	// note that the reference implementation never has withoutMaxSizeCheck == true
+	// checkAvailable doesn't apply
+	if rs.evictionDisabled {
+		return false
+	}
+	return rs.evict()
+}
+
+func (rs *RecordStore) DoExpiration() {
+	// port of: com.hazelcast.internal.nearcache.impl.store.BaseHeapNearCacheRecordStore#doExpiration
+	now := time.Now().UnixMilli()
+	rs.recordsMu.Lock()
+	for k, v := range rs.records {
+		if rs.recordExpired(v, now) {
+			rs.invalidate(k)
+			rs.onExpire()
+		}
+	}
+	rs.recordsMu.Unlock()
+}
+
+func (rs *RecordStore) onEvict(key interface{}, rec *Record, expired bool) {
+	// port of: com.hazelcast.internal.nearcache.impl.store.BaseHeapNearCacheRecordStore#onEvict
+	if expired {
+		rs.incrementExpirations()
+	} else {
+		rs.incrementEvictions()
+	}
+	rs.decrementOwnedEntryCount()
+	rs.decrementOwnedEntryMemoryCost(rs.getTotalStorageMemoryCost(key, rec))
+}
+
+func (rs *RecordStore) tryEvict(candidate evictionCandidate) bool {
+	// port of: com.hazelcast.internal.nearcache.impl.store.HeapNearCacheRecordMap#tryEvict
+	if exists := rs.remove(candidate.Key()); !exists {
+		return false
+	}
+	rs.onEvict(candidate.key, candidate.evictable, false)
+	return true
+}
+
+func (rs *RecordStore) remove(key interface{}) bool {
+	// the key is already in the "made" form, so don't run on rs.makeMapKey on it
+	// assumes rs.recordsMu is locked elsewhere
+	if _, exists := rs.records[key]; !exists {
+		return false
+	}
+	delete(rs.records, key)
+	return true
+}
+
+func (rs *RecordStore) sample(count int) []evictionCandidate {
+	// port of: com.hazelcast.internal.nearcache.impl.store.HeapNearCacheRecordMap#sample
+	// note that count is fixed to 15 in the reference implementation, it is always positive
+	// assumes recordsMu is locked
+	samples := make([]evictionCandidate, count)
+	var idx int
+	for k, v := range rs.records {
+		// access to maps is random
```

Comment:
It's definitely not cryptographically random (neither the Java implementation) but I think it's OK for this purpose. Here is a discussion about that at: https://stackoverflow.com/questions/41019703/is-map-iteration-sufficiently-random-for-randomly-selecting-keys

Slightly modified example on that page: https://go.dev/play/p/gsmgJxH76Gk

The alternative is keeping a separate list of keys and shuffle it which requires more memory or trying to reduce the distribution imbalance by oversampling and choosing the keys using another random method.

Reply:
Thanks a lot for the resources! I think you are right, it gets the job done for now.
The only concern right now is "randomness" of the maps depends on the go version. I think we can either introduce randomness explicitly or just create a test that ensures iterations are somewhat random.

Even if we do not explicitly introduce randomness, I think we need a mechanism to alert us if the behaviour of the maps change, otherwise it would be quite hard to find that bug. Wdyt?

Reply:
To be addressed later